pkg/server/plugin/upstreamauthority/ejbca/ejbca.go

```diff
@@ -0,0 +1,389 @@
+func New() *Plugin {
+	p := &Plugin{}
+	p.hooks.newAuthenticator = p.getAuthenticator
+	p.hooks.getEnv = os.Getenv
+	p.hooks.readFile = os.ReadFile
+	return p
+}
+
+// Configure configures the EJBCA UpstreamAuthority plugin. This is invoked by SPIRE when the plugin is
+// first loaded. After the first invocation, it may be used to reconfigure the plugin.
+func (p *Plugin) Configure(_ context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	config, err := p.parseConfig(req)
+	if err != nil {
+		return nil, err
+	}
+
+	authenticator, err := p.hooks.newAuthenticator(config)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := p.newEjbcaClient(config, authenticator)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to create EJBCA client: %v", err)
+	}
+
+	p.setConfig(config)
+	p.setClient(client)
+	return &configv1.ConfigureResponse{}, nil
+}
+
+// SetLogger is called by the framework when the plugin is loaded and provides
+// the plugin with a logger wired up to SPIRE's logging facilities.
+func (p *Plugin) SetLogger(logger hclog.Logger) {
+	p.logger = logger
+}
+
+// MintX509CAAndSubscribe implements the UpstreamAuthority MintX509CAAndSubscribe RPC. Mints an X.509 CA and responds
+// with the signed X.509 CA certificate chain and upstream X.509 roots. The stream is kept open but new roots will
+// not be published unless the CA is rotated and a new X.509 CA is minted.
+//
+// Implementation note:
+//   - It's important that the EJBCA Certificate Profile and End Entity Profile are properly configured before
+//     using this plugin. The plugin does not attempt to configure these profiles.
+func (p *Plugin) MintX509CAAndSubscribe(req *upstreamauthorityv1.MintX509CARequest, stream upstreamauthorityv1.UpstreamAuthority_MintX509CAAndSubscribeServer) error {
+	var err error
+	if p.client == nil {
+		return status.Error(codes.FailedPrecondition, "ejbca upstreamauthority is not configured")
+	}
+
+	logger := p.logger.Named("MintX509CAAndSubscribe")
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	logger.Debug("Parsing CSR from request")
+	parsedCsr, err := x509.ParseCertificateRequest(req.Csr)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "unable to parse CSR: %v", err)
+	}
+	csrPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: req.Csr})
+
+	logger.Debug("Determining end entity name")
+	endEntityName, err := p.getEndEntityName(config, parsedCsr)
+	if err != nil {
+		return status.Errorf(codes.Internal, "unable to determine end entity name: %v", err)
+	}
+
+	logger.Debug("Preparing EJBCA enrollment request")
+	password, err := generateRandomString(16)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to generate random password: %v", err)
+	}
+	enrollConfig := ejbcaclient.EnrollCertificateRestRequest{}
+	enrollConfig.SetUsername(endEntityName)
+	enrollConfig.SetPassword(password)
```

Comment:
How is this password conveyed with the server?

Reply:
Hi @amartinezfayo!

This password (enrollment code) is conveyed to EJBCA in the `POST /v1/certificate/pkcs10enroll` request and has no direct effect concerning the EJBCA plugin.